### PR TITLE
Cmake: Integratrion

### PR DESCRIPTION
**Pull request by FMMazur**
Just some change to make possible to use with CMake projects.

Needs to implement tests.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,151 @@
+cmake_minimum_required(VERSION 3.8.2)
+project(cglm LANGUAGES C)
+
+set(C_STANDARD 99)
```
Comment (recp) [inline CMakeLists.txt]: We have a `-std=gnu11` flag in **makefile.am** file, MSVC may not support C11/C18... of course I'm not sure. What do you suggest?

Comment (FMMazur) [reply CMakeLists.txt]: I don't have how to test on MSVC, but I'm think cglm project will compile.

**Comment by FMMazur** [comment]
@recp I have create a configuration to possibility to build **tests**, to build with c11, different warning options according to the compilers.

#### Example of Configurations
```bash
$ mkdir build && cd build
$  cmake .. # normal build with c99
$ cmake .. -DCGLM_USE_TEST=ON  # build with tests
$ cmake .. -DCGLM_SHARED=ON # build a shared lib
$ cmake .. -DCGLM_USE_C11 # build with c11
$ cmake .. -DCGLM_SHARED=ON -DCGLM_USE_C11 # build a shared lib with c11
```

**Comment by recp** [comment]
@FMMazur perfect! 

How can we run **cglm** tests?, will `make check` be a enough after build system is generated  by CMake?

**Comment by recp** [comment]
If everything is ok then I'm going to merge this PR.

**Comment by FMMazur** [comment]
I will close this PR for now, any major changes or errors that occur, I will open it again.

**Comment by FMMazur** [comment]
I have that, almost i can, no errors.

**Comment by recp** [comment]
First, great work, thanks!

---

I think  `-DCGLM_SHARED=ON -DCGLM_USE_C11` these options shell be opened as default, if it is possible users can disable it. 

For tests, it is ok to have it with option. But if I run the `make test` (it would be nice to have `make check` or both.), I get this:

Screen Shot 2020-03-27 at 9 18 49 PM

but it would be nice to make it verbose like I did for `make check` (please see it):
Screen Shot 2020-03-27 at 9 19 06 PM

... (scroll to down)

Screen Shot 2020-03-27 at 9 18 54 PM

I must have tested it with a failed case

**Comment by FMMazur** [comment]
I was able to create an alias for `make test` as` make check`, the difference in the code is simple, just a command with the `-V` option, which is a more detailed version of this and allows the test to have an output.

**Comment by recp** [comment]
Is there any issue with `make install` after make build? It seems it doesn't copy the folder structure and overrides files that have same names, or I couldn't do make it.

**Comment by FMMazur** [comment]
> Is there any issue with `make install` after make build? It seems it doesn't copy the folder structure and overrides files that have same names, or I couldn't do make it.

I have added a install path option, this possibility to change install location.
```bash
$ mkdir build && build
$ cmake .. -DCGLM_INSTALL_PATH="<user-path>" # change path
$ make install
```

If don't pass `CGLM_INSTALL_PATH` as argument, will install in [CMAKE_INSTALL_PREFIX](https://cmake.org/cmake/help/v3.13/variable/CMAKE_INSTALL_PREFIX.html#variable:CMAKE_INSTALL_PREFIX)
Its possible pass **absolute** and **relative** path

Created **CGLM_INSTALL_LIBDIR** and **CGLM_INSTALL_INCDIR**, one path to install lib and other to install headers.

**Comment by recp** [comment]
That is good option!

But do you copy sub folders? Did you test it? For instance `cglm/struct/mat4.h` must be placed as it is. It should not override `cglm/mat4.h`

**Comment by FMMazur** [comment]
> That is a good option!
> 
> But do you copy sub folders? Did you test it? For instance `cglm/struct/mat4.h` must be placed as it is. It should not override `cglm/mat4.h`

I fixed this problem, but CMake has an error, and what I was unable to fix when doing it.
**Do not use ~** to refer to the user, instead **use $ HOME**+1, at least on Linux this will be necessary
By example:
```zsh
$ cmake -H. -Bbuild -DCGLM_USE_TEST=ON -DCGLM_INSTALL_PATH="~/Documentos/libs/"  # Looks fine, no problem at all
-- The C compiler identification is GNU 9.1.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /home/MyStorage/Vendor/cglm/build
$ make install -C build
make: Entering directory '/home/MyStorage/Vendor/cglm/build'
make[1]: Entering directory '/home/MyStorage/Vendor/cglm/build'
make[2]: Entering directory '/home/MyStorage/Vendor/cglm/build'
Scanning dependencies of target cglm
make[2]: Leaving directory '/home/MyStorage/Vendor/cglm/build'
make[2]: Entering directory '/home/MyStorage/Vendor/cglm/build'
[  3%] Building C object CMakeFiles/cglm.dir/src/io.c.o
[  6%] Building C object CMakeFiles/cglm.dir/src/box.c.o
[ 10%] Building C object CMakeFiles/cglm.dir/src/ease.c.o
[ 13%] Building C object CMakeFiles/cglm.dir/src/euler.c.o
[ 17%] Building C object CMakeFiles/cglm.dir/src/vec4.c.o
[ 20%] Building C object CMakeFiles/cglm.dir/src/mat2.c.o
[ 24%] Building C object CMakeFiles/cglm.dir/src/mat4.c.o
[ 27%] Building C object CMakeFiles/cglm.dir/src/frustum.c.o
[ 31%] Building C object CMakeFiles/cglm.dir/src/quat.c.o
[ 34%] Building C object CMakeFiles/cglm.dir/src/plane.c.o
[ 37%] Building C object CMakeFiles/cglm.dir/src/project.c.o
[ 41%] Building C object CMakeFiles/cglm.dir/src/bezier.c.o
[ 44%] Building C object CMakeFiles/cglm.dir/src/sphere.c.o
[ 48%] Building C object CMakeFiles/cglm.dir/src/curve.c.o
[ 51%] Building C object CMakeFiles/cglm.dir/src/cam.c.o
[ 55%] Building C object CMakeFiles/cglm.dir/src/vec3.c.o
[ 58%] Building C object CMakeFiles/cglm.dir/src/mat3.c.o
[ 62%] Building C object CMakeFiles/cglm.dir/src/vec2.c.o
[ 65%] Building C object CMakeFiles/cglm.dir/src/affine.c.o
[ 68%] Linking C shared library libcglm.so
make[2]: Leaving directory '/home/MyStorage/Vendor/cglm/build'
[ 68%] Built target cglm
make[2]: Entering directory '/home/MyStorage/Vendor/cglm/build'
Scanning dependencies of target tests
make[2]: Leaving directory '/home/MyStorage/Vendor/cglm/build'
make[2]: Entering directory '/home/MyStorage/Vendor/cglm/build'
# ...
[100%] Linking C executable ../tests
make[2]: Leaving directory '/home/MyStorage/Vendor/cglm/build'
[100%] Built target tests
make[1]: Leaving directory '/home/MyStorage/Vendor/cglm/build'
Install the project...
-- Install configuration: ""
-- Installing: /home/felipe/Documentos/libs/lib/libcglm.so
-- Installing: ~/Documentos/libs/include
CMake Error at cmake_install.cmake:76 (file):
  file INSTALL cannot set permissions on "~/Documentos/libs/include": No such
  file or directory.


make: *** [Makefile:146: install] Error 1
make: Leaving directory '/home/MyStorage/Vendor/cglm/build'
```

### But works with

```zsh
$ cmake -H. -Bbuild -DCGLM_USE_TEST=ON -DCGLM_INSTALL_PATH="$HOME/Documentos/libs/" # builds fine
-- The C compiler identification is GNU 9.1.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /home/MyStorage/Vendor/cglm/build
$ make install -C build
...
Install the project...
-- Install configuration: ""
-- Installing: /home/felipe/Documentos/libs/lib/libcglm.so
-- Up-to-date: /home/felipe/Documentos/libs/include
-- Installing: /home/felipe/Documentos/libs/include/cglm
-- Installing: /home/felipe/Documentos/libs/include/cglm/mat3.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/mat3.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/vec3.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/plane.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/mat2.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/affine.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/quat.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/project.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/sphere.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/curve.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/euler.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/vec3-ext.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/vec2.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/cam.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/vec4-ext.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/mat4.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/color.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/frustum.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/vec2-ext.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/box.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/io.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct/vec4.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/vec3.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/plane.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/mat2.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/affine.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/applesimd.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/quat.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd/arm.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd/avx
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd/avx/affine.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd/avx/mat4.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd/sse2
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd/sse2/mat3.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd/sse2/mat2.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd/sse2/affine.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd/sse2/quat.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd/sse2/mat4.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd/intrin.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd/neon
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd/neon/mat4.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/simd/x86.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/mat3.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/vec3.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/plane.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/mat2.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/affine.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/quat.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/project.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/sphere.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/curve.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/euler.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/vec2.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/cam.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/ease.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/mat4.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/bezier.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/frustum.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/box.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/io.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/call/vec4.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/project.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/sphere.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/curve.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/euler.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/vec3-ext.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/types.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/vec2.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/struct.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/cam.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/ease.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/version.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/vec4-ext.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/mat4.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/color.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/bezier.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/affine-mat.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/frustum.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/util.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/vec2-ext.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/types-struct.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/box.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/common.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/io.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/vec4.h
-- Installing: /home/felipe/Documentos/libs/include/cglm/cglm.h
make: Leaving directory '/home/MyStorage/Vendor/cglm/build'
```

**Comment by recp** [comment]
@FMMazur I think we can remove install path option since **CMake** can provide this option internally. 

For install I could change install path with this:

```zsh
cmake -DCMAKE_INSTALL_PREFIX="~/Downloads/cglm-install-test" ..
```

Result:

```zsh
...
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/sse2/quat.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/arm.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/x86.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/intrin.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/avx
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/avx/mat4.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/avx/affine.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/neon
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/neon/mat4.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/plane.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/call.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/struct.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/curve.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/mat2.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/affine.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/struct
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/struct/vec2.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/struct/box.h
-- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/struct/mat4.h
...
```

it also works with `~`

What do you think? Do we still need install path option?

**Comment by recp** [comment]
@FMMazur ping.

**Comment by FMMazur** [comment]
> @FMMazur I think we can remove install path option since **CMake** can provide this option internally.
> 
> For install, I could change the install path with this:
> 
> ```shell
> cmake -DCMAKE_INSTALL_PREFIX="~/Downloads/cglm-install-test" ..
> make
> make install
> ```
> 
> Result:
> 
> ```shell
> ...
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/sse2/quat.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/arm.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/x86.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/intrin.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/avx
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/avx/mat4.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/avx/affine.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/neon
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/simd/neon/mat4.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/plane.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/call.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/struct.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/curve.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/mat2.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/affine.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/struct
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/struct/vec2.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/struct/box.h
> -- Installing: /Users/recp/Downloads/cglm-install-test/include/cglm/struct/mat4.h
> ...
> ```
> 
> it also works with `~`
> 
> What do you think? Do we still need install path option?
> 
> EDIT:
> 
> To test this, I made a change on `CMakeLists.txt`:
> 
> from
> 
> ```cmake
> INSTALL(DIRECTORY include/ DESTINATION ${CGLM_INSTALL_INCDIR}
> ...
> ```
> 
> to
> 
> ```cmake
> INSTALL(DIRECTORY include/ DESTINATION include
> ...
> ```

I will make that on saturday...

**Comment by recp** [comment]
@FMMazur ping again :)

**Comment by FMMazur** [comment]
I forgot to commit and push... @recp

**Comment by recp** [comment]
> I forgot to commit and push... @recp

☺️

Are we ready to merge this PR then? 

Also, does -O3 flag work with this CMake build configuration? Because I had realized that the size of that built by Raw Autotools build is smaller than the output of the CMake build 🤔

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+endif()
+
+if("${CMAKE_C_COMPILER_ID}" STREQUAL "MSVC")
+  list(APPEND LDFLAGS "/W4")
+elseif("${CMAKE_C_COMPILER_ID}" STREQUAL "GNU")
+  set(C_EXTENSIONS YES)
+  list(APPEND LDFLAGS "-Wall" "-Wpedantic" "-Wstrict-aliasing=2")
+  list(APPEND LDFLAGS "-Werror=strict-prototypes" "-fstrict-aliasing" "-O3")
+elseif("${CMAKE_C_COMPILER_ID}" STREQUAL "CLANG")
+  list(APPEND LDFLAGS "-Wall" "-Weverything" "-Wpedantic")
```
Comment (recp) [inline CMakeLists.txt:30]: the optimization flag `-O3` and `string-aliasing` flag must be included into clang too:

```
"-fstrict-aliasing" "-O3"
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+else(CGLM_STATIC)
+  set(CGLM_BUILD STATIC)
+endif()
+
+if(CGLM_USE_C99)
+  set(C_STANDARD 99)
+endif()
+
+if("${CMAKE_C_COMPILER_ID}" STREQUAL "MSVC")
+  list(APPEND LDFLAGS "/W4")
```
Comment (recp) [inline CMakeLists.txt:24]: We have some flags (optimization flag for instance) on Visuals Studio project which you can find in `win/` folder.

**Comment by realaltffour** [comment]
Hi, everyone!
I have contributed to @FMMazur fork of the project, solving the latest requested [changes](https://github.com/recp/cglm/pull/130#pullrequestreview-391331010)

**Comment by recp** [comment]
@realaltffour thank you very much for helping to this PR. 

Maybe we could merge this PR and get new one to fix or apply the changes? 

cc @FMMazur

**Comment by realaltffour** [comment]
@recp  PR: #138

**Comment by recp** [comment]
@FMMazur thanks for your contributions, this PR is merged 🎉

**Pull request merged**